### PR TITLE
Make node tree order part of the snapshot

### PR DESCRIPTION
**Pull request by ahg-g**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Makes the scheduler's nodeTree object private to the cache and part of the snapshot. The purpose of NodeTree is decide on the iteration order of the nodes (force iterating across zones etc. when evaluating nodes), currently the scheduler iterates over the live tree instead of the snapshot, this has two drawbacks:

1) iterating over the scheduler cache nodeTree requires acquiring a lock, which causes contention (see #72408)
2) causes inconsistencies in clusters that bring up and shutdown nodes frequently (cluster autoscaling)

This PR make the order part of the snapshot, which improves performance (by about 3%) since we don't require locking, and removes the inconsistencies since the scheduler is now iterating over a static list of nodes.

**Which issue(s) this PR fixes**:
Part of #83922

**Does this PR introduce a user-facing change?**:
```release-note
NONE

```

### CODE REVIEW

File: pkg/scheduler/internal/cache/cache.go [hunk 1]
```diff
@@ -716,3 +718,10 @@ func (cache *schedulerCache) GetCSINodeInfo(nodeName string) (*storagev1beta1.CS
 
 	return n, nil
 }
+
+// NumNodes returns the number of nodes.
+func (cache *schedulerCache) NumNodes() int {
```
Comment (liu-cong) [inline pkg/scheduler/internal/cache/cache.go]: It seems that this function is not called anywhere.

File: pkg/scheduler/core/generic_scheduler.go [hunk 2]
```diff
@@ -479,20 +478,22 @@ func (g *genericScheduler) findNodesThatFit(ctx context.Context, state *framewor
 		state.Write(migration.PredicatesStateKey, &migration.PredicatesStateData{Reference: meta})
 
 		checkNode := func(i int) {
-			nodeName := g.cache.NodeTree().Next()
+			nodeInfo := g.nodeInfoSnapshot.NodeInfoList[i]
+			if nodeInfo == nil {
```
Comment (liu-cong) [inline pkg/scheduler/core/generic_scheduler.go]: Can this be nil? Or can we make sure it's not nil?

File: pkg/scheduler/internal/cache/node_tree.go [hunk 3]
```diff
@@ -18,22 +18,21 @@ package cache
...
-
 	"k8s.io/klog"
+	utilnode "k8s.io/kubernetes/pkg/util/node"
 )
 
 // NodeTree is a tree-like data structure that holds node names in each zone. Zone names are
 // keys to "NodeTree.tree" and values of "NodeTree.tree" are arrays of node names.
+// NodeTree is NOT thread-safe, any concurrent updates/reads from it must be synchronized by the caller.
+// It is used only by schedulerCache, and should stay as such.
 type NodeTree struct {
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/internal/cache/node_tree.go]: done

File: pkg/scheduler/internal/cache/cache.go (hunk 1 above)
Comment (ahg-g) [reply pkg/scheduler/internal/cache/cache.go]: done.

File: pkg/scheduler/core/generic_scheduler.go (hunk 2 above)
Comment (ahg-g) [reply pkg/scheduler/core/generic_scheduler.go]: done.

File: pkg/scheduler/internal/cache/cache.go [hunk 4]
```diff
@@ -238,6 +238,14 @@ func (cache *schedulerCache) UpdateNodeInfoSnapshot(nodeSnapshot *schedulernodei
 			}
 		}
 	}
+
+	// Take a snapshot of the nodes order in the tree
+	nodeSnapshot.NodeInfoList = make([]*schedulernodeinfo.NodeInfo, 0, cache.nodeTree.numNodes)
+	for i := 0; i < cache.nodeTree.numNodes; i++ {
+		if n := nodeSnapshot.NodeInfoMap[cache.nodeTree.next()]; n != nil {
```
Comment (liu-cong) [inline pkg/scheduler/internal/cache/cache.go]: Do we know under what circumstances `n` can be nil? Not sure if we should log something if it happens.

Comment (ahg-g) [reply pkg/scheduler/internal/cache/cache.go]: inconsistency between nodeTree and nodeInfoMap, I will log an error.

Comment (ahg-g) [reply pkg/scheduler/internal/cache/cache.go]: done.

File: pkg/scheduler/core/generic_scheduler.go [hunk 5]
```diff
@@ -520,8 +516,8 @@ func (g *genericScheduler) findNodesThatFit(ctx context.Context, state *framewor
 		workqueue.ParallelizeUntil(ctx, 16, int(allNodes), checkNode)
 
 		filtered = filtered[:filteredLen]
-		if len(errs) > 0 {
-			return []*v1.Node{}, FailedPredicateMap{}, framework.NodeToStatusMap{}, errors.CreateAggregateFromMessageCountMap(errs)
+		if err := errCh.ReceiveError(); err != nil {
```
Comment (Huang-Wei) [inline pkg/scheduler/core/generic_scheduler.go]: This doesn't look right, we used to return all predicates errors to the users, but here it changes to a single error.

File: pkg/scheduler/internal/cache/cache.go [hunk 6]
```diff
@@ -238,6 +238,17 @@ func (cache *schedulerCache) UpdateNodeInfoSnapshot(nodeSnapshot *schedulernodei
 			}
 		}
 	}
+
+	// Take a snapshot of the nodes order in the tree
+	nodeSnapshot.NodeInfoList = make([]*schedulernodeinfo.NodeInfo, 0, cache.nodeTree.numNodes)
+	for i := 0; i < cache.nodeTree.numNodes; i++ {
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/cache/cache.go]: The loop seems to be both memory and time consuming, can we have a benchmark testing the whole scheduling cycle?

File: pkg/scheduler/nodeinfo/snapshot.go [hunk 7]
```diff
@@ -20,12 +20,14 @@ import (
...
-// snapshot at the beginning of each scheduling cycle and uses it for its
-// operations in that cycle.
+// Snapshot is a snapshot of cache NodeInfo and NodeTree order. The scheduler takes a
+// snapshot at the beginning of each scheduling cycle and uses it for its operations in that cycle.
 type Snapshot struct {
+	// NodeInfoMap a map of node name to a snapshot of its NodeInfo.
 	NodeInfoMap map[string]*NodeInfo
-	Generation  int64
+	// NodeInfoList is the list of nodes as ordered in the cache's nodeTree.
+	NodeInfoList []*NodeInfo
```
Comment (Huang-Wei) [inline pkg/scheduler/nodeinfo/snapshot.go]: Can we make it `NodeNameList []string`, and then in each `checkNode()`, the "nodeInfo" can be fetched by: `nodeInfo := g.nodeInfoSnapshot. NodeInfoMap[NodeNameList[i]]` (or from a helper method)?

**Review by Huang-Wei — COMMENTED** [review]
Thanks @ahg-g. Some comments below.

**Review by ahg-g — COMMENTED** [review]
Thanks, please see replies.

File: pkg/scheduler/core/generic_scheduler.go (hunk 5 above)
Comment (Huang-Wei) [reply pkg/scheduler/core/generic_scheduler.go]: Ah I misread, `err` is for unexpected internal error, not PredicateFailure. Nevermind then.

File: pkg/scheduler/nodeinfo/snapshot.go (hunk 7 above)
Comment (Huang-Wei) [reply pkg/scheduler/nodeinfo/snapshot.go]: sounds good.

File: pkg/scheduler/core/generic_scheduler.go [hunk 8]
```diff
@@ -479,20 +478,17 @@ func (g *genericScheduler) findNodesThatFit(ctx context.Context, state *framewor
 		state.Write(migration.PredicatesStateKey, &migration.PredicatesStateData{Reference: meta})
 
 		checkNode := func(i int) {
-			nodeName := g.cache.NodeTree().Next()
-
+			nodeInfo := g.nodeInfoSnapshot.NodeInfoList[i]
```
Comment (wojtek-t) [inline pkg/scheduler/core/generic_scheduler.go]: This broke scalability tests: https://github.com/kubernetes/kubernetes/issues/84151

TL;DR; this breaks spreading of pods in large clusters.

What exactly has happened:
1, in large enough clusters, we are using the features of finding only N feasible nodes and scoring only those:
https://github.com/kubernetes/kubernetes/blob/9d173852c14f0e8efb0d67db2c38b8dcfa45b31b/pkg/scheduler/core/generic_scheduler.go#L464

2. Before this change, we were looking for nodes starting at the point where we previously stopped (because Next() was done at the level of the original tree).
3. With this change, we're always starting from 0.

So assume, you have 5k nodes, all are feasible, and numFeasible is choosing 250.
- previously, you first chose nodes [1..250], then [251..500], ... [4751..5000], [1..250], ...
- with this change, we will chose [1..250], [1..250], [1..250], ... until one of those nodes become unfeasible and we will choose something different.

With this PR, next() is called only in UpdateNodeSnapshotInfo:
https://github.com/kubernetes/kubernetes/pull/84014/files#diff-f4a894ca5e905aa5f613269fc967fe2cR206
and if the set of nodes doesn't change, we will pretty much always be generating the same set of nodes.

This kind of breaks the fact that scheduler is scheduling in the whole cluster. While it's not documented feature per-se, I think this isn't the right thing to do.

I'm going to open a revert of this PR to to fix scalability tests (or the half of them, because we seem to have two different regressions), but will wait for your explicit approval.
We can discuss how to fix that later.

Comment (wojtek-t) [reply pkg/scheduler/core/generic_scheduler.go]: heh... it's no longer possible to autorevert it..

Comment (wojtek-t) [reply pkg/scheduler/core/generic_scheduler.go]: Fortunately the conflicts were trivial - opened https://github.com/kubernetes/kubernetes/pull/84222
